File: backend/src/main/java/com/trek/rwanda/destination/Destination.java
package com.trek.rwanda.destination;

import com.trek.rwanda.activity.Activity;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "destinations")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Destination {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(nullable = false, unique = true, length = 250)
    private String slug;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(length = 500)
    private String shortDesc;

    @Column(nullable = false, length = 200)
    private String location;

    @Column(length = 100)
    private String district;

    @Column(length = 100)
    private String province;

    @Column(length = 500)
    private String coverImageUrl;

    @Column(length = 200)
    private String openingHours;

    @Column(columnDefinition = "TEXT")
    private String thingsToKnow;

    @Column(precision = 10, scale = 8)
    private BigDecimal latitude;

    @Column(precision = 11, scale = 8)
    private BigDecimal longitude;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isPublished = false;

    @Builder.Default
    private Integer sortOrder = 0;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "destination_activities",
        joinColumns = @JoinColumn(name = "destination_id"),
        inverseJoinColumns = @JoinColumn(name = "activity_id")
    )
    @Builder.Default
    private Set<Activity> activities = new HashSet<>();

    @OneToMany(mappedBy = "destination", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<DestinationImage> images = new HashSet<>();

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: frontend/src/types/destination.ts
export interface Activity {
  id: string;
  name: string;
  slug: string;
  icon?: string;
}

export interface DestinationImage {
  id: string;
  imageUrl: string;
  altText?: string;
  isCover: boolean;
  sortOrder: number;
}

export interface Destination {
  id: string;
  name: string;
  slug: string;
  description: string;
  shortDesc?: string;
  location: string;
  district?: string;
  province?: string;
  coverImageUrl?: string;
  openingHours?: string;
  thingsToKnow?: string;
  latitude?: number;
  longitude?: number;
  isPublished: boolean;
  isFeatured?: boolean;
  sortOrder: number;
  activities: Activity[];
  images: DestinationImage[];
  createdAt: string;
}

export interface DestinationSummary {
  id: string;
  name: string;
  slug: string;
  shortDesc?: string;
  location: string;
  district?: string;
  coverImageUrl?: string;
  activities: { name: string }[];
}

export interface DestinationDetail extends Destination {
  reviews: {
    averageRating: number;
    totalReviews: number;
  };
  nearbyDestinations: DestinationSummary[];
  safaris: SafariSummary[];
}

export interface SafariSummary {
  id: string;
  name: string;
  slug: string;
  shortDesc?: string;
  durationDays: number;
  price: number;
  currency: string;
  coverImageUrl?: string;
}

export interface Region {
  id: string;
  name: string;
  slug: string;
  description?: string;
  imageUrl?: string;
  sortOrder: number;
}
